### PR TITLE
[Frontend, pytorch] Vc/pytorch lstm

### PR DESCRIPTION
**Pull request by vvchernov**
1. LSTM layer was supported for pytorch frontend (see https://github.com/apache/tvm/issues/6474)
2. Tests for different types of LSTM (unidirectional, bidirectional, stacked, with projection and their combinations) were implemented. Tests are performed for pytorch model and onnx model converted from pytorch

### CODE REVIEW

**Comment by jwfromm** [comment]
@masahi can you help review and shepherd this PR?

**Comment by vvchernov** [comment]
Hello @masahi, I have finished my work and waiting for your review and response

File: tests/python/frontend/pytorch/test_lstms.py [hunk 1]
```diff
@@ -0,0 +1,382 @@
...
+        bidirectional=False,
+        proj_size=0,
+        rnd_weights_init=False,
+    ):
+        super().__init__()
+
+        self.device = device
+        self.batch_first = batch_first
+
+        # Network defition
```
Comment (emijiayw) [inline tests/python/frontend/pytorch/test_lstms.py:74]: Network definition

**Comment by masahi** [comment]
thanks @vvchernov this is great. Please fix the lint issue.

**Comment by vvchernov** [comment]
@masahi @jwfromm Hello guys, I have fixed all issues related to lint and resolved merge conflicts, but tvm-ci failed. After that I chose main with success status in jenkins (this one https://ci.tlcpack.ai/blue/organizations/jenkins/tvm/detail/main/1292/pipeline) and merged it to my branch. but tvm-ci failed again. In both cases it seems that it is not related to my changings in the code. First Build:GPU failed, currently python3: i386 failed. What do you think about it? May be you can give me a hash with stable commit for main branch

**Comment by masahi** [comment]
@vvchernov Both failures look new to me, they can be flaky tests. Can you try running CI again (rebase and push)?

**Comment by vvchernov** [comment]
@masahi I followed your recommendation and tvm-ci passed successfully. But I observed that my test did not check case where LSTM layer without bias weights. I updated test and checked these cases with combination of other LSTM modifications. Observed bugs were fixed (please review last commit). It seems that all modification of LSTM supported by pytorch are supported on tvm side.

File: tests/python/frontend/pytorch/test_lstms.py (hunk 1 above)
Comment (masahi) [reply tests/python/frontend/pytorch/test_lstms.py:76]: yeah let's fix this @vvchernov

File: tests/python/frontend/pytorch/test_lstms.py [hunk 2]
```diff
@@ -0,0 +1,421 @@
...
+def compare(input, gold_data, epsilon=1e-6):
+    remain = np.abs(gold_data - input)
+    err = np.max(remain)
+    if err < epsilon:
+        print("SUCCESS: RESULTS ARE THE SAME WITH MAX ERROR {} AND EPSILON {}".format(err, epsilon))
+    else:
+        print("WARNING: RESULTS ARE NOT THE SAME WITH ERROR {}".format(err))
+
+
+if __name__ == "__main__":
```
Comment (masahi) [inline tests/python/frontend/pytorch/test_lstms.py:179]: The CI uses pytest, which picks up only functions names which start with `test_`. So it is likely that CI is not running this test right now. Please wrap this main test code with `test_` function.

File: tests/python/frontend/pytorch/test_lstms.py [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+
+
+def compare(input, gold_data, rtol=1e-5, atol=1e-5):
+    tvm.testing.assert_allclose(input, gold_data, rtol=rtol, atol=atol)
+    # remain = np.abs(gold_data - input)
+    # err = np.max(remain)
+    # if err < 1e-6:
+    #     print("SUCCESS: RESULTS ARE THE SAME WITH MAX ERROR {} AND EPSILON {}".format(err, 1e-6))
+    # else:
+    #     print("WARNING: RESULTS ARE NOT THE SAME WITH ERROR {}".format(err))
```
Comment (masahi) [inline tests/python/frontend/pytorch/test_lstms.py:181]: Remove these comments

File: tests/python/frontend/pytorch/test_lstms.py [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+                        # Import model to Relay
+                        shape_list = [("input", input_shape)]
+                        mod, params = relay.frontend.from_pytorch(traced_script_module, shape_list)
+
+                        # Model compilation by tvm
+                        target = tvm.target.Target("llvm", host="llvm")
+                        dev = tvm.cpu(0)
+                        with tvm.transform.PassContext(opt_level=3):
+                            lib = relay.build(mod, target=target, params=params)
+                    elif format == "onnx":
```
Comment (masahi) [inline tests/python/frontend/pytorch/test_lstms.py:303]: Remove onnx test

File: tests/python/frontend/pytorch/test_lstms.py [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+    #     print("SUCCESS: RESULTS ARE THE SAME WITH MAX ERROR {} AND EPSILON {}".format(err, 1e-6))
+    # else:
+    #     print("WARNING: RESULTS ARE NOT THE SAME WITH ERROR {}".format(err))
+
+
+def check_lstm_with_type(lstm_type):
+    # Create outdir directory to keep temporal files
+    out_dir = Path.cwd().joinpath("output")
+    out_dir.mkdir(exist_ok=True, parents=True)
+    has_proj = "p" in lstm_type
```
Comment (masahi) [inline tests/python/frontend/pytorch/test_lstms.py:188]: No need to create temp directory if you remove onnx test.

File: tests/python/frontend/pytorch/test_lstms.py [hunk 6]
```diff
@@ -0,0 +1,378 @@
...
+                    )
+                    # Execute
+                    m.run()
+                    # Get outputs (converted to numpy array)
+                    tvm_output = m.get_output(0).numpy()
+
+                    compare(tvm_output, golden_output_batch)
+
+    # Remove output directory with tmp files
+    shutil.rmtree(out_dir)
```
Comment (masahi) [inline tests/python/frontend/pytorch/test_lstms.py:363]: Remove it with onnx test

File: tests/python/frontend/pytorch/test_lstms.py [hunk 7]
```diff
@@ -0,0 +1,378 @@
...
+                        traced_script_module = torch.jit.trace(model, dummy_input).eval()
+
+                        # Import model to Relay
+                        shape_list = [("input", input_shape)]
+                        mod, params = relay.frontend.from_pytorch(traced_script_module, shape_list)
+
+                        # Model compilation by tvm
+                        target = tvm.target.Target("llvm", host="llvm")
+                        dev = tvm.cpu(0)
+                        with tvm.transform.PassContext(opt_level=3):
```
Comment (masahi) [inline tests/python/frontend/pytorch/test_lstms.py:301]: Please test on cuda too.

File: tests/python/frontend/pytorch/test_lstms.py [hunk 8]
```diff
@@ -0,0 +1,378 @@
...
+                golden_output_batch = model.forward(dummy_input.to(device)).detach().cpu().numpy()
+
+                dtype = "float32"
+                h_zeros = np.zeros(input_hidden_shape, dtype=dtype)
+                if has_proj:
+                    h_zeros = np.zeros(input_hidden_shape_with_proj, dtype=dtype)
+                c_zeros = np.zeros(input_hidden_shape, dtype=dtype)
+
+                tvm_output = None
+                for format in ("ts", "onnx"):
```
Comment (masahi) [inline tests/python/frontend/pytorch/test_lstms.py:289]: format

File: tests/python/frontend/pytorch/test_lstms.py (hunk 4 above)
Comment (vvchernov) [reply tests/python/frontend/pytorch/test_lstms.py:303]: My idea to check modifications of LSTM layer converted from pytorch to onnx is related to that pytorch frontend supports less ML layers than onnx one. It means that complicated neural network contained LSTM can be not supported from pytorch model on tvm side, but onnx model converted from pytorch can be. Moreover there is specific of stacked LSTM support, I do not think that it is tested on onnx frontend side

Comment (masahi) [reply tests/python/frontend/pytorch/test_lstms.py:303]: Ok. Then please address my other comments. Also you can avoid creating a temp directory by writing onnx file to an in-memory buffer, see https://github.com/facebookresearch/detr/blob/master/test_all.py#L136-L140

**Comment by masahi** [comment]
@vvchernov Don't forget https://github.com/apache/tvm/pull/8447#discussion_r668195327

File: tests/python/frontend/pytorch/test_lstms.py (hunk 1 above)
Comment (vvchernov) [reply tests/python/frontend/pytorch/test_lstms.py]: done

File: tests/python/frontend/pytorch/test_lstms.py (hunk 3 above)
Comment (vvchernov) [reply tests/python/frontend/pytorch/test_lstms.py]: done

File: tests/python/frontend/pytorch/test_lstms.py (hunk 5 above)
Comment (vvchernov) [reply tests/python/frontend/pytorch/test_lstms.py:177]: done

File: tests/python/frontend/pytorch/test_lstms.py (hunk 6 above)
Comment (vvchernov) [reply tests/python/frontend/pytorch/test_lstms.py]: done

File: tests/python/frontend/pytorch/test_lstms.py (hunk 4 above)
Comment (vvchernov) [reply tests/python/frontend/pytorch/test_lstms.py:292]: I does not work for onnx model loading, but I found correct way with io.BytesIO using. done

File: tests/python/frontend/pytorch/test_lstms.py (hunk 7 above)
Comment (vvchernov) [reply tests/python/frontend/pytorch/test_lstms.py:290]: done

**Review by masahi — APPROVED** [review]

**Comment by masahi** [comment]
Thanks @vvchernov !

**Comment by masahi** [comment]
@vvchernov I was informed that the test added in this PR is taking more than 20 min of CI time. Can you reduce the number of test cases?

**Comment by cheimu** [comment]
Hi there. I found this error still exist. I just clone this repo yesterday
image

**Comment by vvchernov** [comment]
Hello @cheimu! Dropout is not supported on TVM side. Therefore LSTM with dropout does not work. In general it is so because some people think that dropout is used for training only, but I know that for some TTS model it can be used for inference. Unfortunately the current state is that dropout op is dummy, it has attributes but does not have compute implementation. I do not know when dropout will be supported on TVM side.
